### ISSUE TITLE
Plater needs a new script/mod event: Uninitialization

### ISSUE THREAD

**Issue by Bananaman** [Tercioo/Plater-Nameplates#105]
- Initialization: Triggers when the mod first starts up, and every time the user disables/re-enables the mod.

- Constructor: Triggers when nameplates first appear on the screen.

- Destructor: When the user disables the mod, this triggers for all visible plates. **If no plates are visible, this event doesn't trigger at all.**

There is, from what I can see, no real way to do a guaranteed "this code will run when the user disables my addon" in Plater.

The destructor is not reliable, since it doesn't run *at all* if no plates are visible.

Therefore, it prevents programmers from creating Plater mods that do complex setup and teardowns. In my case, I'm hooking into another addon, leeching its data, presenting it in Plater instead. If the user disables my Plater mod, I need to restore the original 3rd party addon's behavior. But I can't do that, because there is no reliable event for it.

I even tried going through all of Plater's source code, looking for a way to loop through the list of compiled and loaded mods and checking which ones are enabled. That is kind of doable, but extremely hacky.

What would really improve things here is if Plater has a new event:

- Uninitialization: Triggers whenever the mod is disabled (whether via auto-unload from load conditions, or via the user manually disabling the mod).

**Comment by Bananaman** [comment]
This is related to the "Mod Destructor's modTable is nil" bug:

https://github.com/Tercioo/Plater-Nameplates/issues/104

Proposed Plater "mod disabled" workflow for a mod:

1. Trigger destructors for all visible frames, with modTable provided.
2. Trigger uninitialization script for the mod, with modTable provided.
3. Erase the modTable.

**Comment by cont1nuity** [comment]
Agreed, makes sense.

**Comment by cont1nuity** [comment]
Please test against the next alpha version.

**Comment by Bananaman** [comment]
@cont1nuity Thank you so much. I am testing the Deinitialization script now and it's working perfectly. It receives the modTable and I'm able to unhook from the 3rd party addon in there, I love it so much! ;-)

**Comment by cont1nuity** [comment]
Thank you fort testing. If you find any issues, please let me know.

**Comment by Bananaman** [comment]
I will continue testing as I port my mod to these new functions. But it seems perfect!

**Comment by Bananaman** [comment]
@cont1nuity Still testing, and everything still seems perfect!

I forgot to say that this message seems a bit unclear:

`["Deinitialization"] = "Executed once for the mod when it is unloaded. Used to de-initialize the global mod environment 'modTable' and the mod for the last time.",`

It makes it sound like deinitialization only runs once, ever. And that it only runs when the mod is being unloaded for the last time.

From what I can see, it always runs every time the mod is unloaded (whether automatically via load-conditions or manually unloaded).

I notice, however, that the "execute once" language is used in initializer too: `["Initialization"] = "Executed once for the mod when it is compiled. Used to initialize the global mod environment 'modTable'.",`

Perhaps something like this is clearer then?

`["Initialization"] = "Executed once for the mod every time it is compiled. Used to initialize the global mod environment 'modTable'.",`

`["Deinitialization"] = "Executed once for the mod every time it is unloaded. Used to de-initialize the global mod environment 'modTable' and the mod.",`

As a programmer reading the docs, those descriptions would be a bit clearer. I've had to do a lot of manual testing prior to this too, until I figured out when the different hooks run. 😄 So if the descriptions can become a bit clearer, that would be a win.
